# Practice (Decorator cont.)

Q) Write a Python program that implements a decorator to enforce rate limits on a function.

In [11]:
import time
def rate_limits(max_calls,period):
    def decorator(func):
        calls=0
        last_Reset=time.time()

        def wrapper(*args,**kargs):
            nonlocal calls,last_Reset
            elapsed = time.time() - last_Reset
            if elapsed > period:
                calls=0
                last_Reset=time.time()
            if calls >=max_calls:
                raise Exception("Rate Limit exceeded,please try again later.")
            calls+=1
            return func(*args,**kargs)
        return wrapper
    return decorator

@rate_limits(max_calls=6,period=10)
def call():
    print("API call executed successfully..")

for i in range(8):
    try:
        call()
    except Exception as e:
        print(f"Error occurred: {e}")

API call executed successfully..
API call executed successfully..
API call executed successfully..
API call executed successfully..
API call executed successfully..
API call executed successfully..
Error occurred: Rate Limit exceeded,please try again later.
Error occurred: Rate Limit exceeded,please try again later.


Q) Python program that implements a decorator to enforce type checking on the arguments of a function.

In [15]:
import inspect
def checker(func):
    def wrapper(*args,**kwargs):
        sign=inspect.signature(func)
        para=sign.parameters
        for i,arg in enumerate(args):
            para_name=list(para.keys())[i]
            para_type=para[para_name].annotation
            if not isinstance(arg,para_type):
                raise TypeError(f"Argument '{para_name}' must be of type '{para_type.__name__}'")
        for para_name,arg in kwargs.items():
            para_type=para[para_name].annotation
            if not isinstance(arg,para_type):
                raise TypeError(f"Argument '{para_name}' must be of type '{para_type.__name__}'")
        return func(*args,**kwargs)
    return wrapper

@checker
def multi(x:int,y:int)-> int:
    return x*y

r=multi(3,4)
print(f"Result: {r}")

r= multi("3",7)

Result: 12


TypeError: Argument 'x' must be of type 'int'

Q) Python program that implements a decorator to measure the memory usage of a function.

In [20]:
import tracemalloc
def memory(func):
    def wrapper(*args,**kwargs):
        tracemalloc.start()
        r=func(*args,**kwargs)
        snapshot=tracemalloc.take_snapshot()
        top_stats=snapshot.statistics("lineno")
        print(f"Memory usage of {func.__name__}")
        for stat in top_stats[:5]:
            print(stat)
        return r
    return wrapper


@memory
def cal_fact(n):
    if n==0:
        return 1
    else:
        return n*cal_fact(n-1)
r=cal_fact(5)
print(f"Factorial:",r)

Memory usage of cal_fact
/usr/lib/python3.10/ast.py:50: size=489 KiB, count=8925, average=56 B
/usr/lib/python3.10/genericpath.py:77: size=224 KiB, count=2285, average=100 B
/home/trellis/.local/lib/python3.10/site-packages/asttokens/line_numbers.py:47: size=140 KiB, count=748, average=192 B
/home/trellis/.local/lib/python3.10/site-packages/pygments/lexer.py:504: size=85.9 KiB, count=1100, average=80 B
/home/trellis/.local/lib/python3.10/site-packages/executing/executing.py:169: size=75.9 KiB, count=196, average=396 B
Memory usage of cal_fact
/usr/lib/python3.10/ast.py:50: size=489 KiB, count=8925, average=56 B
/usr/lib/python3.10/genericpath.py:77: size=179 KiB, count=1636, average=112 B
/home/trellis/.local/lib/python3.10/site-packages/asttokens/line_numbers.py:47: size=140 KiB, count=748, average=192 B
/home/trellis/.local/lib/python3.10/site-packages/executing/executing.py:169: size=75.9 KiB, count=196, average=396 B
/home/trellis/.local/lib/python3.10/site-packages/executing/execu

Q) Python program that implements a decorator to provide caching with expiration time for a function.

In [25]:
import time

def cache_with_expiry(expiration_time):
    cache = {}

    def decorator(func):
        def wrapper(*args, **kwargs):
            current_time = time.time()
            for key in list(cache.keys()):
                if current_time - cache[key]['time'] >= expiration_time:
                    del cache[key]
            key = (func.__name__, args, frozenset(kwargs.items()))
            if key in cache:
                print(f"Result retrieved from cache for {func.__name__}.")
                return cache[key]['value']
            else:
                result = func(*args, **kwargs)
                cache[key] = {'value': result, 'time': time.time()}
                return result
        return wrapper
    return decorator

@cache_with_expiry(expiration_time=5)
def expensive_function(n):
    print("Executing expensive function...")
    time.sleep(2)  
    return n * n

# Call the decorated function
print(expensive_function(5))  
print(expensive_function(5))  
time.sleep(2)
print(expensive_function(5))  


Executing expensive function...
25
Result retrieved from cache for expensive_function.
25
Result retrieved from cache for expensive_function.
25


# Practice (Programming puzzles)

Q) Python program that accepts a list of integers and calculates the length and the fifth element. Return true if the length of the list is 8 and the fifth element occurs thrice in the said list.

In [34]:
def check(x):
    if len(x) >= 8:
        if x.count(x[4]) == 3:
            return True
        else:
            return False
    return False

a=[i for i in input("Enter numbers space separated:").split(" ")]
print(check(a))

True


Q) Python program that accepts an integer and determines whether it is greater than 4^4 and which is 4 mod 34.

In [39]:
def check(x):
    if x > pow(4,4) and x%34 == 4:
            return True
    return False

a=int(input("Enter number:"))
print(check(a))

True


Q) We are making n stone piles! The first pile has n stones. If n is even, then all piles have an even number of stones. If n is odd, all piles have an odd number of stones. Each pile must more stones than the previous pile but as few as possible. Write a Python program to find the number of stones in each pile.<br>
Input: 2<br>
Output:<br>
[2, 4]<br>
Input: 10<br>
Output:<br>
[10, 12, 14, 16, 18, 20, 22, 24, 26, 28]

In [43]:
def stones(n):
    return [n + 2 * i for i in range(n)]

n=int(input("Enter number for piles:"))
print(stones(n))

[10, 12, 14, 16, 18, 20, 22, 24, 26, 28]


Q) Write a Python program to check the nth-1 string is a proper substring of the nth string in a given list of strings.<br>
Input:<br>
['a', 'abb', 'sfs', 'oo', 'de', 'sfde']<br>
Output:<br>
True<br>
Input:<br>
['a', 'abb', 'sfs', 'oo', 'ee', 'sfde']<br>
Output:<br>
False<br>

In [53]:
import re

def substring(x):
    last=x[-1]
    sec_last=x[-2]
    if re.search(sec_last,last) == None:
        return False
    else:
        return True

a=[i for i in input("Enter strings space separated:").split(" ")]
print(substring(a))

True


Q) Write a Python program to test a list of one hundred integers between 0 and 999, which all differ by ten from one another. Return True otherwise False.<br>
Input:<br>
[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 390, 400, 410, 420, 430, 440, 450, 460, 470, 480, 490, 500, 510, 520, 530, 540, 550, 560, 570, 580, 590, 600, 610, 620, 630, 640, 650, 660, 670, 680, 690, 700, 710, 720, 730, 740, 750, 760, 770, 780, 790, 800, 810, 820, 830, 840, 850, 860, 870, 880, 890, 900, 910, 920, 930, 940, 950, 960, 970, 980, 990]<br>
Output:<br>
True<br>
Input:<br>
[0, 20, 40, 60, 80, 100, 120, 140, 160, 180, 200, 220, 240, 260, 280, 300, 320, 340, 360, 380, 400, 420, 440, 460, 480, 500, 520, 540, 560, 580, 600, 620, 640, 660, 680, 700, 720, 740, 760, 780, 800, 820, 840, 860, 880, 900, 920, 940, 960, 980]<br>
Output:<br>
False<br>

In [55]:
def test(li):
    return all(i in range(1000) and abs(i - j) >= 10 for i in li for j in li if i != j) and len(set(li)) == 100

nums = list(range(0, 1000, 10))
print("Original list:")
print(nums)

print("Check whether the said list contains one hundred integers between 0 and 999 which all differ by ten from one another:")
print(test(nums))
nums = list(range(0, 1000, 20))
print("\nOriginal list:")
print(nums)
print("Check whether the said list contains one hundred integers between 0 and 999 which all differ by ten from one another:")
print(test(nums))


Original list:
[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 390, 400, 410, 420, 430, 440, 450, 460, 470, 480, 490, 500, 510, 520, 530, 540, 550, 560, 570, 580, 590, 600, 610, 620, 630, 640, 650, 660, 670, 680, 690, 700, 710, 720, 730, 740, 750, 760, 770, 780, 790, 800, 810, 820, 830, 840, 850, 860, 870, 880, 890, 900, 910, 920, 930, 940, 950, 960, 970, 980, 990]
Check whether the said list contains one hundred integers between 0 and 999 which all differ by ten from one another:
True

Original list:
[0, 20, 40, 60, 80, 100, 120, 140, 160, 180, 200, 220, 240, 260, 280, 300, 320, 340, 360, 380, 400, 420, 440, 460, 480, 500, 520, 540, 560, 580, 600, 620, 640, 660, 680, 700, 720, 740, 760, 780, 800, 820, 840, 860, 880, 900, 920, 940, 960, 980]
Check whether the said list contains one hundred integers between 0 and 999 which all differ by ten from one an

Q) Write a Python program to check a given list of integers where the sum of the first i integers is i.<br>
Input:<br>
[0, 1, 2, 3, 4, 5]<br>
Output:<br>
False<br>
Input:<br>
[1, 1, 1, 1, 1, 1]<br>
Output:<br>
True

In [61]:
def test(li, i):
    return sum(li[:i]) == i

nums = [int(x) for x in input("Enter integers separated by spaces: ").split()]
i = int(input("Enter the value of 'i': "))
print("Original list:")
print(nums)
print("Check the said list, where the sum of the first i integers is i: i =", i)
print(test(nums, i))


Original list:
[0, 1, 2, 3, 4, 5, 6]
Check the said list, where the sum of the first i integers is i: i = 3
True


Q) Write a Python program to split a string of words separated by commas and spaces into two lists, words and separators.<br>
Input: W3resource Python, Exercises.<br>
Output:<br>
[['W3resource', 'Python', 'Exercises.'], [' ', ', ']]<br>

In [62]:
def test(string):
    import re
    merged = re.split(r"([ ,]+)", string)
    return [merged[::2], merged[1::2]]

s = input("Enter a string: ")
print("Original string:", s)
print("Split the said string into 2 lists: words and separators:")
print(test(s))


Original string: koza,mod
Split the said string into 2 lists: words and separators:
[['koza', 'mod'], [',']]


Q) Write a Python program to find a list of integers containing exactly four distinct values, such that no integer repeats twice consecutively among the first twenty entries.

In [66]:
def twice(x):
    return all([x[i] != x[i+1] for i in range(len(x)-1 )]) and len(set(x)) == 4

nums = [int(x) for x in input("Enter integers separated by spaces: ").split()]
print(twice(nums))

True


Q) Given a string consisting of whitespace and groups of matched parentheses, write a Python program to split it into groups of perfectly matched parentheses without any whitespace.<br>
Input:<br>
( ()) ((()()())) (()) ()<br>
Output:<br>
['(())', '((()()()))', '(())', '()']<br>
Input:<br>
() (( ( )() ( )) ) ( ())<br>
Output:<br>
['()', '((()()()))', '(())']<br>

In [75]:
def test(combined):
    ls = []
    s2 = ""
    for s in combined.replace(' ', ''):
        s2 += s
        if s2.count("(") == s2.count(")"):
            ls.append(s2)
            s2 = ""
    return ls

combined = input("Enter a string with parentheses: ")
print(f"Parentheses string: {combined}")
print(f"Separate parentheses groups of the said string:\n{test(combined)}")

Parentheses string: () (( ( )() (  )) ) ( ())
Separate parentheses groups of the said string:
['()', '((()()()))', '(())']


Q) Write a Python program to find the indexes of numbers in a given list below a given threshold.<br>
Original list:<br>
[0, 12, 45, 3, 4923, 322, 105, 29, 15, 39, 55]<br>
Threshold: 100<br>
Check the indexes of numbers of the said list below the given threshold:<br>
[0, 1, 2, 3, 7, 8, 9, 10]<br>
Original list:<br>
[0, 12, 4, 3, 49, 9, 1, 5, 3]<br>
Threshold: 10<br>
Check the indexes of numbers of the said list below the given threshold:<br>
[0, 2, 3, 5, 6, 7, 8]<br>

In [79]:
def find_indexes_below_threshold(lst, threshold):
    return [index for index, num in enumerate(lst) if num < threshold]

# Take input for the list of numbers
nums = [int(x) for x in input("Enter list of numbers separated by space: ").split()]

# Take input for the threshold
threshold = int(input("Enter the threshold: "))

result = find_indexes_below_threshold(nums, threshold)
print("Indexes of numbers below the threshold:", result)


Indexes of numbers below the threshold: [1, 3, 4, 6]


In [2]:
def anagram(x,y):
    if sorted(a.lower()) == sorted(b.lower()):
        return("THe entered strings are anagram:")
    else:
        return("THe entered strings are not anagram:")

a=input("ENter first word:")
b=input("Enter second word:")
print(anagram(a,b))

THe entered strings are not anagram:
